hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DefaultSource.scala

```diff
@@ -301,11 +302,29 @@ object DefaultSource {
               resolveBaseFileOnlyRelation(sqlContext, globPaths, userSchema, metaClient, parameters)
             }
           case (COPY_ON_WRITE, QUERY_TYPE_INCREMENTAL_OPT_VAL, _) =>
-            if (useNewParquetFileFormat) {
-              new HoodieCopyOnWriteIncrementalHadoopFsRelationFactory(
-                sqlContext, metaClient, parameters, userSchema, isBootstrappedTable).build()
+            if (SparkConfigUtils.containsConfigProperty(parameters, INCREMENTAL_READ_VERSION)) {
+              val writeTableVersion = Integer.parseInt(parameters(INCREMENTAL_READ_VERSION.key))
+              if (writeTableVersion >= 8) {
+                if (useNewParquetFileFormat) {
+                  new HoodieCopyOnWriteIncrementalHadoopFsRelationFactory(
+                    sqlContext, metaClient, parameters, userSchema, isBootstrappedTable).build()
+                } else {
+                  new IncrementalRelationV2(sqlContext, parameters, userSchema, metaClient, RangeType.CLOSED_CLOSED)
+                }
+              } else {
+                new IncrementalRelationV1(sqlContext, parameters, userSchema, metaClient)
+              }
             } else {
-              new IncrementalRelation(sqlContext, parameters, userSchema, metaClient)
+              if (metaClient.getTableConfig.getTableVersion.versionCode() >= 8) {
```

Comment:
can we simplify all this to just a single if-else switch. Looks like we just need a couple more boolean flags based on whether its get or not.. 

```
val tableVersion = if (SparkConfigUtils.containsConfigProperty(parameters, INCREMENTAL_READ_VERSION) {
             Integer.parseInt(parameters(INCREMENTAL_READ_VERSION.key) 
        } else {
             metaClient.getTableConfig.getTableVersion
        }
        
 ```